src/seeds.proposals.cpp

```diff
@@ -249,6 +262,91 @@ void proposals::calcvotepow() {
 
 }
 
+ACTION proposals::migvotepow(uint64_t cycle) {
+  require_auth(_self);
+
+  auto pitr = props.find(131);// NOTE migration method from cycle 32 onward
+  uint64_t all_total = 0;
+  uint64_t cmp_total = 0;
+  uint64_t cmp_num = 0;
+  uint64_t all_num = 0;
+  while( pitr != props.end() && pitr -> passed_cycle == cycle) {
+    name prop_type = get_type(pitr->fund);
+    if (prop_type == alliance_type) {
+      all_total += pitr -> total;
+      all_num += 1;
+    } else {
+      cmp_total += pitr -> total; 
+      cmp_num += 1;
+    }
+    pitr++;
+  }
+
+  set_support_level(cycle, cmp_num, cmp_total, campaign_type);
+  set_support_level(cycle, all_num, all_total, alliance_type);
+
+  print("cycle " +
+    std::to_string(cycle) +
+    " alliance props: " + 
+    std::to_string(all_num) +
+    " alliance total votes: " + 
+    std::to_string(all_total) +
+
+    " campaign props: " + 
+    std::to_string(cmp_num) +
+    " campaign total votes: " + 
+    std::to_string(cmp_total) 
+  );
+
+}
+
+void proposals::set_support_level(uint64_t cycle, uint64_t num_proposals, uint64_t votes_cast, name type) {
+  
+  uint64_t votes_needed = calc_voice_needed(votes_cast, num_proposals);
+
+  support_level_tables support(get_self(), type.value);
+  auto sitr = support.find(cycle);
+  
+  if (sitr != support.end()) {
+    support.modify(sitr, _self, [&](auto & item){
+      item.num_proposals = num_proposals;
+      item.total_voice_cast = votes_cast;
+      item.voice_needed = votes_needed;
+    });
+  } else {
+    support.emplace(_self, [&](auto & item){
+      item.propcycle = cycle;
+      item.num_proposals = num_proposals;
+      item.total_voice_cast = votes_cast;
+      item.voice_needed = calc_voice_needed(votes_cast, num_proposals);
+    });
+  }
+}
+
+void proposals::add_voice_cast(uint64_t cycle, uint64_t voice_cast, name type) {
+  support_level_tables support(get_self(), type.value);
+  auto sitr = support.find(cycle);
+  check(sitr != support.end(), "cycle not found "+std::to_string(cycle));
+  support.modify(sitr, _self, [&](auto & item){
+    uint64_t total_voice = item.total_voice_cast + voice_cast;
+    item.total_voice_cast = total_voice;
+    item.voice_needed = calc_voice_needed(total_voice, item.num_proposals);
+  });
+}
+
+uint64_t proposals::calc_voice_needed(uint64_t total_voice, uint64_t num_proposals) {
+  return ceil(total_voice * (get_quorum(num_proposals) / 100.0));
```

Comment:
There's one small bugfix here - using ceil() of this value

Example if 19.2 votes are required at minimum, then really, 20 votes are required, as 19 < 19.2 and 20 > 19.2